jsonpickle/ext/numpy.py

```diff
@@ -58,18 +65,175 @@ class NumpyNDArrayHandler(NumpyBaseHandler):
+    """
+
+    size_treshold = 16
+    compression = zlib
+
+    def flatten(self, obj, data):
+        """encode numpy to json"""
+        if self.size_treshold > obj.size or self.size_treshold is None:
+            # store as json
+            data = super(NumpyNDArrayHandlerBinary, self).flatten(obj, data)
+        else:
+            # store as binary
+            assert obj.data.contiguous
+            self.flatten_dtype(obj.dtype, data)
+            buffer = obj.tobytes(order=None)
+            if self.compression:
+                buffer = self.compression.compress(buffer)
+            values = jsonpickle.util.b64encode(buffer)
+            data['values'] = self.context.flatten(values, reset=False)
+            data['shape'] = obj.shape
+            if obj.flags.f_contiguous:
+                data['order'] = 'F'
+            if obj.flags.writeable is False:
+                data['writeable'] = False
+            data['byteorder'] = native_byteorder
         return data
 
     def restore(self, data):
-        dtype = self.restore_dtype(data)
-        return np.array(self.context.restore(data['values'], reset=False),
-                        dtype=dtype)
+        """decode numpy from json"""
+        values = data['values']
+        if isinstance(values, list):
+            # decode json
+            arr = super(NumpyNDArrayHandlerBinary, self).restore(data)
+        else:
+            # decode binary representation
+            values = self.context.restore(values, reset=False)
+            buffer = jsonpickle.util.b64decode(values)
+            if self.compression:
+                buffer = self.compression.decompress(buffer)
+            arr = np.ndarray(
+                buffer=buffer,
+                dtype=self.restore_dtype(data).newbyteorder(data['byteorder']),
+                shape=data.get('shape'),
+                order=data.get('order', 'C')
+            ).copy() # make a copy, to force the result to own the data
+            if not data.get('writeable', True):
+                arr.flags.writeable = False
+
+        return arr
+
+
+class NumpyNDArrayHandlerView(NumpyNDArrayHandlerBinary):
+    """Pickles references inside ndarrays, or array-views
+
+    Notes
+    -----
+    The current implementation has some restrictions.
+
+    'base' arrays, or arrays which are viewed by other arrays, must be f-or-c-contiguous.
+    This is not such a large restriction in practice, because all numpy array creation is c-contiguous by default.
+    Releasing this restriction would be nice though; especially if it can be done without bloating the design too much.
+
+    Furthermore, ndarrays which are views of array-like objects implementing __array_interface__,
+    but which are not themselves nd-arrays, are deepcopied with a warning,
+    as we cannot guarantee whatever custom logic such classes implement is correctly reproduced.
+    """
+    mode = 'warn'
```

Comment:
I guess the choice is between allowing to register instances, or adding a mechanism for specifying constructor args to specific handlers.  It does seem like allowing for instances would be the simplest and most versatile design without needing to do too much plumbing for handler-specific arguments.